Modules/_pickle.c

```diff
@@ -1601,15 +1601,15 @@ memo_get(PicklerObject *self, PyObject *key)
 /* Store an object in the memo, assign it a new unique ID based on the number
    of objects currently stored in the memo and generate a PUT opcode. */
 static int
-memo_put(PicklerObject *self, PyObject *obj)
+memo_put(PicklerObject *self, PyObject *obj, int opt)
 {
     char pdata[30];
     Py_ssize_t len;
     Py_ssize_t idx;
 
     const char memoize_op = MEMOIZE;
 
-    if (self->fast)
+    if (self->fast || (opt && Py_REFCNT(obj) == 1))
```

Comment:
Requiring `Py_REFCNT(obj)` to be 1 pretty strong right? Does this only affect objects created using the `C API`?

Reply:
It mostly affects temporary objects. For example, if `__reduce__` returns `constructor, ((x, y),)` then the tuple `(x, y)` will not be memoized. This is a case of namedtuples.

Another example: if you have a list of unique numbers, strings, tuples, etc, then items of the list will not be memoized as the only reference to the item is from the list.